[{"text": "We're using Tomcat in our product. Under traffic we can reproduce the following exceptions:\n\nNov 23, 2015 CST 08:34:58.115 PM org.apache.coyote.http11.Http11NioProcessor\nINFO: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.\n\nNov 23, 2015 CST 08:34:58.118 PM org.apache.coyote.http11.Http11NioProcessor\nSEVERE: Error finishing response\n\njava.lang.NullPointerException\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.flushBuffer(InternalNioOutputBuffer.java:234)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.addToBB(InternalNioOutputBuffer.java:189)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.commit(InternalNioOutputBuffer.java:177)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:741)\n\tat org.apache.coyote.Response.action(Response.java:177)\n\tat org.apache.coyote.http11.AbstractOutputBuffer.endRequest(AbstractOutputBuffer.java:369)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.endRequest(AbstractHttp11Processor.java:1804)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1143)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:673)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\nWe have several applications deployed on our Tomcat. Most of the applications are simple Jersey Rest service applications. Above NullPointer is related to one of them (basically, the one which is called during traffic - I bet other apps will have the same issue if they would be called). Also we have one application which is using WebSockets (Tomcat implementation) and Async support for Comet. WebSocket interface is used during traffic testing while Comet is NOT (Servlets are not called). We have the following in our Tomcat main web.xml file (comparing to the default one):\n\n\n    <filter>\n        <filter-name>CorsFilter</filter-name>\n        <filter-class>com.avaya.acc.base.web.filters.CorsFilter</filter-class>\n        <async-supported>true</async-supported>\n    </filter>\n    <filter-mapping>\n        <filter-name>CorsFilter</filter-name>\n        <url-pattern>/*</url-pattern>\n    </filter-mapping>\n\n    <filter>\n        <filter-name>NoCachingFilter</filter-name>\n        <filter-class>com.avaya.acc.base.web.filters.NoCachingFilter</filter-class>\n        <async-supported>true</async-supported>\n        <init-param>\n          <param-name>Set-Expires</param-name>\n          <param-value>false</param-value>\n        </init-param>\n        <init-param>\n          <param-name>No-Store</param-name>\n          <param-value>true</param-value>\n        </init-param>\n    </filter>\n    <filter-mapping>\n        <filter-name>NoCachingFilter</filter-name>\n        <url-pattern>/*</url-pattern>\n    </filter-mapping>\n    \n    <!-- enable caching for static content -->\n    <context-param>\n        <param-name>NoCachingFilter enableCaching</param-name>\n        <param-value>Set-Expires=true;No-Store=false;patterns=*.js,*.swf,*.jar,*.png,*.jpg,*.css</param-value>\n    </context-param>\n    \n    <!-- Declaring this param in web.xml for specific webapp doesn't work, so leaving it here -->\n    <context-param>\n        <param-name>stinger.excluded.url.patterns</param-name>\n        <param-value>/Comet/StartLiveSession,/Comet/Data,/Comet/EndLiveSession</param-value>\n    </context-param>\n \nAnd the following in the web.xml for application which has issues:\n\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\n<web-app version=\"2.5\" xmlns=\"http://java.sun.com/xml/ns/javaee\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_2_5.xsd\">\n    <servlet>\n        <servlet-name>Live Viewer</servlet-name>\n        <servlet-class>org.glassfish.jersey.servlet.ServletContainer</servlet-class>\n\n        <init-param>\n                <param-name>javax.ws.rs.Application</param-name>\n                <param-value>com.avaya.acc.wcs.api.viewer.rest.app.LiveViewerRestApp</param-value>\n        </init-param>\n\n        <init-param>\n            <!-- integrates JERSEY and JACKSON -->\n            <param-name>com.sun.jersey.api.json.POJOMappingFeature</param-name>\n            <param-value>true</param-value>\n        </init-param>\n        <init-param>\n            <param-name>sendfileSize</param-name>\n            <param-value>-1</param-value>\n        </init-param>\n        <load-on-startup>1</load-on-startup>\n    </servlet>\n\n    <servlet-mapping>\n        <servlet-name>Live Viewer</servlet-name>\n        <url-pattern>/rest/*</url-pattern>\n    </servlet-mapping>\n</web-app>\n\nI tried to remove async-supported from all our web.xml files and the issue is gone. So it's definitely related to <async-supported> configuration - without even actually using async functionality. We still need async-supported for our Comet interface so just disabling async is not an option for us.\n\nI can turn on some additional logging and reproduce again if this will help - just let me know which logging is needed.", "tags": [], "creator": "anmalyshev@mail.ru", "attachment_id": null, "count": 0, "id": 186523, "time": "2015-11-24T21:39:08Z", "bug_id": 58646, "creation_time": "2015-11-24T21:39:08Z", "is_private": false}, {"count": 1, "attachment_id": null, "bug_id": 58646, "is_private": false, "id": 186528, "time": "2015-11-24T22:26:59Z", "creator": "anmalyshev@mail.ru", "creation_time": "2015-11-24T22:26:59Z", "tags": [], "text": "After some more testing it turns out it's not related to async-supported tag. It's just intermittent. Also, sometimes this one happens before below exceptions:\n\nNov 24, 2015 CST 04:18:48.480 PM org.apache.catalina.connector.CoyoteAdapter\nINFO: Encountered a non-recycled request and recycled it forcedly.\n\norg.apache.catalina.connector.CoyoteAdapter$RecycleRequiredException\n\tat org.apache.catalina.connector.CoyoteAdapter.checkRecycled(CoyoteAdapter.java:710)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.recycle(AbstractHttp11Processor.java:1828)\n\tat org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.release(Http11NioProtocol.java:219)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:727)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)"}, {"count": 2, "tags": [], "creator": "anmalyshev@mail.ru", "is_private": false, "text": "Note also that the amount of regular HTTP request during testing is not huge - nearly 10-12 requests/second. Amount of websocket traffic is huge, but it does not seem to affect these exceptions - issue can happen in the beginning of the test when WS traffic is still small as well as in the end of the test.", "id": 186529, "time": "2015-11-24T22:37:06Z", "bug_id": 58646, "creation_time": "2015-11-24T22:37:06Z", "attachment_id": null}, {"count": 3, "tags": [], "text": "1)\n> Nov 23, 2015 CST 08:34:58.115 PM org.apache.coyote.http11.Http11NioProcessor\n> INFO: Error parsing HTTP request header\n> Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.\n\nThe above logging is controlled by org.apache.juli.logging.UserDataHelper.CONFIG system property [1].\n\nThe log message was likely written by \"catch (Throwable t)\" clause in AbstractHttp11Processor.process(). That is the only place where UserDataHelper is used in a processor.\n\nUnfortunately, it does not log the stacktrace, even with debug logging. It can be improved with a code change in AbstractHttp11Processor.java line 1058,\n\n-                            getLog().debug(message);\n+                            getLog().debug(message, t);\n\n(An easy test case: I have seen such messages when erroneously trying to connect with https:// to a http connector.)\n\n\n2)\nI do not see thread names in your log. It is likely that the two errors are related (being just 3ms apart), but it is unclear whether the two errors occurred in the same thread.\n\nIt is known that org.apache.juli.OneLineFormatter writes the thread names.\n\n\n3) Usual advice with misbehaving requests/responses is to add the following property into conf/catalina.properties file:  [2]\n\norg.apache.catalina.connector.RECYCLE_FACADES=true\n\nIt should make your configuration more robust against common mistake of accessing Request/Response objects outside of their life cycle.\n\n\n4) BTW, Tomcat 8.0.29 was released several hours ago. I think that it will be announced tomorrow.\n\n\n[1] http://tomcat.apache.org/tomcat-8.0-doc/config/systemprops.html#Logging\n[2] http://tomcat.apache.org/tomcat-8.0-doc/security-howto.html#System_Properties", "attachment_id": null, "id": 186532, "creator": "knst.kolinko@gmail.com", "time": "2015-11-24T22:53:53Z", "bug_id": 58646, "creation_time": "2015-11-24T22:53:53Z", "is_private": false}, {"count": 4, "tags": [], "creator": "markt@apache.org", "attachment_id": null, "text": "Correct title based on additional information.", "id": 186549, "time": "2015-11-25T14:10:20Z", "bug_id": 58646, "creation_time": "2015-11-25T14:10:20Z", "is_private": false}, {"count": 5, "tags": [], "bug_id": 58646, "attachment_id": null, "id": 186550, "time": "2015-11-25T14:19:12Z", "creator": "markt@apache.org", "creation_time": "2015-11-25T14:19:12Z", "is_private": false, "text": "I've added the stack trace to the log message.\n\nInitial impressions of this issue are an application bug along the lines of retaining a reference to the response."}, {"count": 6, "tags": [], "bug_id": 58646, "text": "I don't think it's application bug. During traffic testing only the following requests are sent:\n\n1) Requests to get static content. Fully processed by tomcat - only filters apply.\n2) Request to get one Jersey REST service which is really simple and is not using request or response directly.\n3) WebSocket requests processed by Tomcat WS implementation.\n\nThe only custom part here are filters, but they are pretty simple and I can disable them and run test again. Also, they're not remembering request or response objects - only get and use them inside doFilter() method.\n\nNote though that setting org.apache.catalina.connector.RECYCLE_FACADES=true seem to fix an issue (I can't be sure as before issue was intermittent as well). But at least I had a few traffic runs without this issue.", "id": 186563, "time": "2015-11-25T18:01:34Z", "creator": "anmalyshev@mail.ru", "creation_time": "2015-11-25T18:01:34Z", "is_private": false, "attachment_id": null}, {"count": 7, "attachment_id": null, "bug_id": 58646, "text": "Ok, It seems like the issue happens even with org.apache.catalina.connector.RECYCLE_FACADES=true.\n\nFollowing are the exceptions I got:\n\nNov 25, 2015 CST 06:10:52.115 PM http-nio-135.60.87.69-443-exec-4 org.apache.coyote.http11.Http11NioProcessor\nINFO: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.\n\n---- 500 milliseconds between these 2 -------\n\nNov 25, 2015 CST 06:10:52.697 PM http-nio-135.60.87.69-443-exec-5 org.apache.catalina.core.ContainerBase.[Tomcat].[localhost].[/].[jsp]\nSEVERE: Servlet.service() for servlet [jsp] in context with path [] threw exception [java.lang.IllegalStateException: The request object has been recycled and is no longer associated with this facade] with root cause\n\njava.lang.IllegalStateException: The request object has been recycled and is no longer associated with this facade\n\tat org.apache.catalina.connector.RequestFacade.getMethod(RequestFacade.java:743)\n\tat org.apache.jsp.index_jsp._jspService(index_jsp.java:82)\n\tat org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:70)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:729)\n\tat org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:438)\n\tat org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:396)\n\tat org.apache.jasper.servlet.JspServlet.service(JspServlet.java:340)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:729)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)\n\tat org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)\n\tat com.avaya.acc.base.web.filters.NoCachingFilter.doFilter(NoCachingFilter.java:106)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)\n\tat com.avaya.acc.base.web.filters.CorsFilter.handleSimpleCORS(CorsFilter.java:438)\n\tat com.avaya.acc.base.web.filters.CorsFilter.doFilter(CorsFilter.java:289)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:217)\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)\n\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)\n\tat com.avaya.acc.base.web.valves.StingerValve.invoke(StingerValve.java:150)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)\n\tat com.avaya.acc.base.tomcat.MCPTomcatAccessLogValve.invoke(MCPTomcatAccessLogValve.java:290)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:673)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\nNov 25, 2015 CST 06:10:52.703 PM http-nio-135.60.87.69-443-exec-7 org.apache.catalina.core.ContainerBase.[Tomcat].[localhost].[/].[jsp]\nSEVERE: Servlet.service() for servlet [jsp] in context with path [] threw exception\n\njava.lang.NullPointerException\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:246)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:217)\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)\n\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)\n\tat com.avaya.acc.base.web.valves.StingerValve.invoke(StingerValve.java:150)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)\n\tat com.avaya.acc.base.tomcat.MCPTomcatAccessLogValve.invoke(MCPTomcatAccessLogValve.java:290)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:673)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\nNov 25, 2015 CST 06:10:52.704 PM http-nio-135.60.87.69-443-exec-7 org.apache.coyote.http11.Http11NioProcessor\nSEVERE: Error processing request\n\njava.lang.NullPointerException\n\tat org.apache.catalina.connector.Request.notifyAttributeAssigned(Request.java:1500)\n\tat org.apache.catalina.connector.Request.setAttribute(Request.java:1490)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:159)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)\n\tat com.avaya.acc.base.tomcat.MCPTomcatAccessLogValve.invoke(MCPTomcatAccessLogValve.java:290)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:673)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\nNov 25, 2015 CST 06:10:52.705 PM http-nio-135.60.87.69-443-exec-7 org.apache.coyote.http11.Http11NioProcessor\nSEVERE: Error finishing response\n\njava.lang.NullPointerException\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.flushBuffer(InternalNioOutputBuffer.java:234)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.addToBB(InternalNioOutputBuffer.java:189)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.commit(InternalNioOutputBuffer.java:177)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:741)\n\tat org.apache.coyote.Response.action(Response.java:177)\n\tat org.apache.coyote.http11.AbstractOutputBuffer.endRequest(AbstractOutputBuffer.java:369)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.endRequest(AbstractHttp11Processor.java:1804)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1143)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:673)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\nNote that 3 different threads are involved.\n\nI'll collect access logs and enable org.apache.juli.logging.UserDataHelper.CONFIG=INFO next time.", "id": 186572, "time": "2015-11-26T01:00:55Z", "creator": "anmalyshev@mail.ru", "creation_time": "2015-11-26T01:00:55Z", "tags": [], "is_private": false}, {"count": 8, "tags": [], "bug_id": 58646, "text": "(In reply to Alexander Malyshev from comment #6)\n> I don't think it's application bug.\n\nAll the evidence so far points to an application bug being more likely than a Tomcat bug.\n\n> The only custom part here are filters,\n\nThat statement is not correct as a quick look at the stack traces shows:\ncom.avaya.acc.base.tomcat.MCPTomcatAccessLogValve\ncom.avaya.acc.base.web.valves.StingerValve\n\n> Note though that setting org.apache.catalina.connector.RECYCLE_FACADES=true\n> seem to fix an issue\n\nIt won't fix the issue, it will just make it more obvious where the problem lies by triggering NPEs.", "id": 186657, "time": "2015-11-30T09:50:41Z", "creator": "markt@apache.org", "creation_time": "2015-11-30T09:50:41Z", "is_private": false, "attachment_id": null}, {"count": 9, "tags": [], "creator": "anmalyshev@mail.ru", "attachment_id": null, "is_private": false, "id": 186689, "time": "2015-11-30T20:16:32Z", "bug_id": 58646, "creation_time": "2015-11-30T20:16:32Z", "text": "Well, I don't see how NullPointer in Tomcat code can possibly be an application bug.\n\nI forgot about these two valves, but, again, they never store request/response objects so they definitely are not using these objects outside of invoke method - and it's legal to use them inside of this method.\n\nI added some additional logging (as well as upgraded to 8.0.29) and now waiting to reproduce it again. As usually - once I enabled logging it's gone :) \n\nI always saw this one:\n\nNov 23, 2015 CST 08:34:58.115 PM org.apache.coyote.http11.Http11NioProcessor\nINFO: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.\n\nbefore all the null pointers so I'm 99% sure these are related. Let me get a stack trace of this one as well as additional logging and we'll see what's happening."}, {"count": 10, "tags": [], "text": "(In reply to Alexander Malyshev from comment #9)\n> Well, I don't see how NullPointer in Tomcat code can possibly be an\n> application bug.\n\npackage com.your.code;\n\npublic class BrokeFilter implements Filter {\n  public void toFilter(ServletRequest request, ServletResponse response, FilterChain chain) {\n    chain.doFilter(null, null);\n  }\n}\n\nTomcat is going to throw an NPE somewhere down the line, yet the bug is in application code.\n\nYes, Tomcat could protect against this kind of thing, but adding thousands of null-checks to things that obviously shouldn't be null isn't a good use of time in a high-performance server.", "attachment_id": null, "bug_id": 58646, "id": 186695, "time": "2015-11-30T23:29:26Z", "creator": "chris@christopherschultz.net", "creation_time": "2015-11-30T23:29:26Z", "is_private": false}, {"count": 11, "tags": [], "creator": "markt@apache.org", "attachment_id": null, "id": 186950, "time": "2015-12-14T21:38:05Z", "bug_id": 58646, "creation_time": "2015-12-14T21:38:05Z", "is_private": false, "text": "There isn't enough information here (yet) to demonstrate a Tomcat bug or reproduce the issue."}, {"count": 12, "tags": [], "creator": "anmalyshev@mail.ru", "attachment_id": null, "is_private": false, "id": 186953, "time": "2015-12-15T04:53:50Z", "bug_id": 58646, "creation_time": "2015-12-15T04:53:50Z", "text": "Yes, I agree. For some reason I can easily reproduce it with clean tomcat 8.0.29, but once I add logging to it to investigate further - it disappears..\nI'll be running more traffic during this week so, hopefully, I'll be able to catch it."}, {"count": 13, "tags": [], "text": "(In reply to Alexander Malyshev from comment #12)\n> Yes, I agree. For some reason I can easily reproduce it with clean tomcat\n> 8.0.29, but once I add logging to it to investigate further - it disappears..\n> I'll be running more traffic during this week so, hopefully, I'll be able to\n> catch it.\n\nWere you able to reproduce the issue with traffic ?", "attachment_id": null, "id": 187488, "creator": "Juliesur@gmail.com", "time": "2016-01-07T18:22:02Z", "bug_id": 58646, "creation_time": "2016-01-07T18:22:02Z", "is_private": false}, {"count": 14, "tags": [], "creator": "Juliesur@gmail.com", "attachment_id": null, "id": 187490, "time": "2016-01-07T18:24:48Z", "bug_id": 58646, "creation_time": "2016-01-07T18:24:48Z", "is_private": false, "text": "I am seeing similar errors in my log under traffic. I am using tomcat 8.0.15"}, {"count": 15, "tags": [], "text": "Ok, finally I was able to reproduce it with logs.\nI associated ID with each request object and I added logging to AbstractHttp11Processor  class to process() and recycle() methods.\n\nIn process() method I added the following log in the very beginning:\n\n    getLog().info(\"Starting to process request with id \" + this.request.id + \". Calling method is:\\n\" + java.lang.Thread.currentThread().getStackTrace()[3]);\n\nand the following log later then request path is parsed (copying a few above lines as well to make it clean):\n\n        if ((!getInputBuffer().parseRequestLine(this.keptAlive)) && \n          (handleIncompleteRequestLineRead()))\n        {\n          break;\n        }\n\n        getLog().info(\"Processing \" + this.request.method() + \" request to \" + this.request.requestURI() + \".\");\n\nIn the recycle() method I added log in the very beginning:\n\ngetLog().info(\"Recycling processor with id \" + this.request.id + \". Calling method is:\\n\" + java.lang.Thread.currentThread().getStackTrace()[2]);\n    \nWhat I got is that 2 different Tomcat threads started to process the same request (logs are ordered properly):\n\nJan 15, 2016 MSK 04:52:13.914 PM http-nio-135.60.87.68-443-exec-23 org.apache.coyote.http11.Http11NioProcessor\nINFO: Starting to process request with id 13. Calling method is:\norg.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\nJan 15, 2016 MSK 04:52:13.914 PM http-nio-135.60.87.68-443-exec-23 org.apache.coyote.http11.Http11NioProcessor\nINFO: Processing POST request to /viewer/rest/live_viewer/.\n\nJan 15, 2016 MSK 04:52:13.914 PM http-nio-135.60.87.68-443-exec-17 org.apache.coyote.http11.Http11NioProcessor\nINFO: Starting to process request with id 13. Calling method is:\norg.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\nJan 15, 2016 MSK 04:52:13.914 PM http-nio-135.60.87.68-443-exec-17 org.apache.coyote.http11.Http11NioProcessor\nINFO: Processing POST request to /viewer/rest/live_viewer/.\n\nJan 15, 2016 MSK 04:52:13.914 PM http-nio-135.60.87.68-443-exec-17 org.apache.coyote.http11.Http11NioProcessor\nINFO: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.\n\njava.lang.IllegalStateException: Unexpected state: headers already parsed. Buffer not recycled?\n\tat org.apache.coyote.http11.AbstractNioInputBuffer.parseHeaders(AbstractNioInputBuffer.java:375)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1029)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\nJan 15, 2016 MSK 04:52:13.914 PM http-nio-135.60.87.68-443-exec-23 com.avaya.acc.wcs.logic.api.GetPresentationTocCommand\nINFO: Processing LIVE_VIEWER event for client with Participant ID null\n\n\nLast log is aour app log and means that http-nio-135.60.87.68-443-exec-23 thread processed request successfully.\nI think there is a synchronization issue somethere in tomcat code which makes it possible for 2 threads to process the same request object simultaneously.\n\nHere is another exception - but the reason seem to be the same:\n\nJan 15, 2016 MSK 04:52:14.028 PM http-nio-135.60.87.68-443-exec-16 org.apache.coyote.http11.Http11NioProcessor\nINFO: Starting to process request with id 13. Calling method is:\norg.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\nJan 15, 2016 MSK 04:52:14.028 PM http-nio-135.60.87.68-443-exec-16 org.apache.coyote.http11.Http11NioProcessor\nINFO: Processing GET request to /wcsws/WebSocket.\n\nJan 15, 2016 MSK 04:52:14.028 PM http-nio-135.60.87.68-443-exec-1 org.apache.coyote.http11.Http11NioProcessor\nINFO: Starting to process request with id 13. Calling method is:\norg.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\nJan 15, 2016 MSK 04:52:14.029 PM http-nio-135.60.87.68-443-exec-1 org.apache.coyote.http11.Http11NioProcessor\nINFO: Processing GET request to /wcsws/WebSocket.\n\nJan 15, 2016 MSK 04:52:14.029 PM http-nio-135.60.87.68-443-exec-1 org.apache.coyote.http11.Http11NioProcessor\nINFO: Recycling processor with id 13. Calling method is:\norg.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.release(Http11NioProtocol.java:218)\n\nJan 15, 2016 MSK 04:52:14.029 PM http-nio-135.60.87.68-443-exec-16 org.apache.catalina.core.ContainerBase.[Tomcat].[localhost].[/wcsws].[default]\nSEVERE: Servlet.service() for servlet [default] in context with path [/wcsws] threw exception\n\njava.lang.NullPointerException\n\tat org.apache.catalina.connector.Request.notifyAttributeAssigned(Request.java:1505)\n\tat org.apache.catalina.connector.Request.access$000(Request.java:118)\n\tat org.apache.catalina.connector.Request$3.set(Request.java:3380)\n\tat org.apache.catalina.connector.Request.setAttribute(Request.java:1462)\n\tat org.apache.catalina.connector.RequestFacade.setAttribute(RequestFacade.java:539)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:272)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)\n\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)\n\tat com.avaya.acc.base.web.valves.StingerValve.invoke(StingerValve.java:150)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)\n\tat com.avaya.acc.base.tomcat.MCPTomcatAccessLogValve.invoke(MCPTomcatAccessLogValve.java:269)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1100)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\nJan 15, 2016 MSK 04:52:14.040 PM http-nio-135.60.87.68-443-exec-16 org.apache.coyote.http11.Http11NioProcessor\nSEVERE: Error processing request\n\njava.lang.NullPointerException\n\tat org.apache.catalina.connector.Request.notifyAttributeAssigned(Request.java:1505)\n\tat org.apache.catalina.connector.Request.setAttribute(Request.java:1495)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:158)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)\n\tat com.avaya.acc.base.tomcat.MCPTomcatAccessLogValve.invoke(MCPTomcatAccessLogValve.java:269)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1100)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\nJan 15, 2016 MSK 04:52:14.040 PM http-nio-135.60.87.68-443-exec-16 org.apache.coyote.http11.Http11NioProcessor\nSEVERE: Error finishing response\n\njava.lang.NullPointerException\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.flushBuffer(InternalNioOutputBuffer.java:236)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.addToBB(InternalNioOutputBuffer.java:191)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.commit(InternalNioOutputBuffer.java:179)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:741)\n\tat org.apache.coyote.Response.action(Response.java:177)\n\tat org.apache.coyote.http11.AbstractOutputBuffer.endRequest(AbstractOutputBuffer.java:369)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.endRequest(AbstractHttp11Processor.java:1813)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1152)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)", "is_private": false, "id": 187709, "creator": "anmalyshev@mail.ru", "time": "2016-01-15T14:26:45Z", "bug_id": 58646, "creation_time": "2016-01-15T14:26:45Z", "attachment_id": null}, {"count": 16, "tags": [], "creator": "anmalyshev@mail.ru", "attachment_id": null, "id": 187796, "time": "2016-01-18T12:07:16Z", "bug_id": 58646, "creation_time": "2016-01-18T12:07:16Z", "is_private": false, "text": "I can add more logs and reproduce again - just let me know the place to add logs."}, {"count": 17, "attachment_id": null, "bug_id": 58646, "is_private": false, "id": 187806, "time": "2016-01-18T16:59:03Z", "creator": "markt@apache.org", "creation_time": "2016-01-18T16:59:03Z", "tags": [], "text": "Thanks for the additional information. It is certainly more information of the right type but it isn't (yet) enough to figure out what is going on.\n\nI think whatever is going wrong, has already gone wrong by the time the log messages provided in comment #15 are reached. What might provide further insight is including the log messages for at least one whole (correctly processed) request prior to the point where things went wrong for each of the threads involved and the processor involved.\n\nIn terms of what might be going wrong, one possibility is that a Processor is getting added more than once to the recycled processor pool. I can'tsee how that might happen but adding logging in AbstractProtocol.RecycledProcessors.push() inside the if (offer) { ... } block might provide an idea of if something is going wrong there."}, {"count": 18, "tags": [], "creator": "iztok.cergol@gmail.com", "attachment_id": null, "is_private": false, "id": 187986, "time": "2016-01-27T12:12:49Z", "bug_id": 58646, "creation_time": "2016-01-27T12:12:49Z", "text": "I have a very similar situation (on Tomcat 7.0.65) that happens after several hours of service:\n\ngen 12, 2016 7:03:19 PM org.apache.coyote.http11.AbstractHttp11Processor process\nInformazioni: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.\njava.lang.NumberFormatException: null\n        at java.lang.Integer.parseInt(Integer.java:454)\n        at java.lang.Integer.parseInt(Integer.java:527)\n        at org.myapp.web.servlet.StreamingServlet.doStreaming(StreamingServlet.java:138)\n        at org.myapp.web.servlet.StreamingServlet.service(StreamingServlet.java:43)\n        at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)\n        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)\n        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)\n        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n        at org.myapp.web.filter.LogRequestFilter.doFilter(LogRequestFilter.java:117)\n        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)\n        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)\n        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)\n        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)\n        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)\n        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:956)\n        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)\n        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:423)\n        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1079)\n        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1757)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1716)\n        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n        at java.lang.Thread.run(Thread.java:744)\ngen 12, 2016 7:03:19 PM org.apache.catalina.connector.CoyoteAdapter checkRecycled\nInformazioni: Encountered a non-recycled response and recycled it forcedly.\norg.apache.catalina.connector.CoyoteAdapter$RecycleRequiredException\n        at org.apache.catalina.connector.CoyoteAdapter.checkRecycled(CoyoteAdapter.java:592)\n        at org.apache.coyote.http11.AbstractHttp11Processor.recycle(AbstractHttp11Processor.java:1813)\n        at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.release(Http11NioProtocol.java:218)\n        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:708)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1757)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1716)\n        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n        at java.lang.Thread.run(Thread.java:744)\n\ngen 12, 2016 7:03:27 PM org.apache.coyote.http11.AbstractHttp11Processor process\nInformazioni: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.\njava.lang.NumberFormatException: null\n        at java.lang.Integer.parseInt(Integer.java:454)\n        at java.lang.Integer.parseInt(Integer.java:527)\n        at org.myapp.web.servlet.StreamingServlet.doStreaming(StreamingServlet.java:138)\n        at org.myapp.web.servlet.StreamingServlet.service(StreamingServlet.java:43)\n        at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)\n        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)\n        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)\n        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n        at org.myapp.web.filter.LogRequestFilter.doFilter(LogRequestFilter.java:117)\n        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)\n        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)\n        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)\n        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)\n        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)\n        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:956)\n        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)\n        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:423)\n\t\t\ngen 12, 2016 7:03:27 PM org.apache.catalina.connector.CoyoteAdapter checkRecycled\nInformazioni: Encountered a non-recycled response and recycled it forcedly.\norg.apache.catalina.connector.CoyoteAdapter$RecycleRequiredException\n        at org.apache.catalina.connector.CoyoteAdapter.checkRecycled(CoyoteAdapter.java:592)\n        at org.apache.coyote.http11.AbstractHttp11Processor.recycle(AbstractHttp11Processor.java:1813)\n        at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.release(Http11NioProtocol.java:218)\n        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:708)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1757)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1716)\n        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n        at java.lang.Thread.run(Thread.java:744)\n\n        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1079)\n        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1757)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1716)\n        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n        at java.lang.Thread.run(Thread.java:744)\ngen 12, 2016 7:03:27 PM org.apache.coyote.http11.AbstractHttp11Processor process\nGrave: Error processing request\njava.lang.NullPointerException\n        at org.apache.coyote.http11.InternalNioOutputBuffer.flushBuffer(InternalNioOutputBuffer.java:233)\n        at org.apache.coyote.http11.InternalNioOutputBuffer.endRequest(InternalNioOutputBuffer.java:121)\n        at org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:762)\n        at org.apache.coyote.Response.action(Response.java:174)\n        at org.apache.coyote.Response.finish(Response.java:274)\n        at org.apache.catalina.connector.OutputBuffer.close(OutputBuffer.java:319)\n        at org.apache.catalina.connector.Response.finishResponse(Response.java:553)\n        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:450)\n        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1079)\n        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1757)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1716)\n        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n        at java.lang.Thread.run(Thread.java:744)\n\ngen 12, 2016 7:03:27 PM org.apache.coyote.http11.AbstractHttp11Processor endRequest\nGrave: Error finishing response\njava.lang.NullPointerException\n        at org.apache.coyote.http11.InternalNioOutputBuffer.flushBuffer(InternalNioOutputBuffer.java:233)\n        at org.apache.coyote.http11.InternalNioOutputBuffer.endRequest(InternalNioOutputBuffer.java:121)\n        at org.apache.coyote.http11.AbstractHttp11Processor.endRequest(AbstractHttp11Processor.java:1788)\n        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1131)\n        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1757)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1716)\n        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n        at java.lang.Thread.run(Thread.java:744)\n\ngen 12, 2016 7:03:33 PM org.apache.catalina.connector.CoyoteAdapter checkRecycled\nInformazioni: Encountered a non-recycled response and recycled it forcedly.\norg.apache.catalina.connector.CoyoteAdapter$RecycleRequiredException\n        at org.apache.catalina.connector.CoyoteAdapter.checkRecycled(CoyoteAdapter.java:592)\n        at org.apache.coyote.http11.AbstractHttp11Processor.recycle(AbstractHttp11Processor.java:1813)\n        at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.release(Http11NioProtocol.java:218)\n        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:708)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1757)\n        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1716)\n        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n        at java.lang.Thread.run(Thread.java:744)\n\n\n\nBelow you can see the localhost_access_log. Pay attention on lines with date '01/Jan/1970:00:59:59'. The request has no method, no URL, but Tomcat still return code 200 and no bytes. In other cases returns 500.\n\t\t\n\t\t\n\t\t\n10.121.210.247 user1 [12/Jan/2016:19:03:19 +0100] \"GET /streaming?id=147366 HTTP/1.1\" 400 0 8206\n- - [01/Jan/1970:00:59:59 +0100] \"-\" 200 0 -\n10.121.210.247 user2 [12/Jan/2016:19:03:19 +0100] \"GET /streaming?id=145255 HTTP/1.1\" 200 149 24605\n10.120.56.102 user1 [12/Jan/2016:19:03:22 +0100] \"GET /streaming?id=147366 HTTP/1.1\" 200 233 24605\n10.121.210.247 user2 [12/Jan/2016:19:03:22 +0100] \"GET /streaming?id=145255 HTTP/1.1\" 200 155 24605\n10.120.56.102 user1 [12/Jan/2016:19:03:26 +0100] \"GET /streaming?id=147366 HTTP/1.1\" 200 244 24605\n10.121.210.247 user2 [12/Jan/2016:19:03:27 +0100] \"GET /streaming?id=145258 HTTP/1.1\" 400 0 8206\n10.121.210.247 - [01/Jan/1970:00:59:59 +0100] \"-\" 200 0 8194\n- - [01/Jan/1970:00:59:59 +0100] \"-\" 500 0 -\n10.121.210.247 user2 [12/Jan/2016:19:03:28 +0100] \"GET /streaming?id=145258 HTTP/1.1\" 200 153 24605\n10.121.210.247 user2 [12/Jan/2016:19:03:30 +0100] \"GET /streaming?id=145258 HTTP/1.1\" 200 153 24605\n10.120.56.102 user1 [12/Jan/2016:19:03:31 +0100] \"GET /streaming?id=147366 HTTP/1.1\" 200 214 24605\n10.120.56.102 user1 [12/Jan/2016:19:03:33 +0100] \"GET /streaming?id=147366 HTTP/1.1\" 200 255 24605\n10.121.210.247 user2 [12/Jan/2016:19:03:33 +0100] \"GET /streaming?id=145258 HTTP/1.1\" 400 0 16406\n10.121.210.247 - [01/Jan/1970:00:59:59 +0100] \"-\" 200 0 -\n10.120.56.102 user1 [12/Jan/2016:19:03:34 +0100] \"GET /streaming?id=147366 HTTP/1.1\" 200 266 24605"}, {"count": 19, "tags": [], "bug_id": 58646, "attachment_id": null, "id": 188096, "time": "2016-02-01T08:42:24Z", "creator": "markt@apache.org", "creation_time": "2016-02-01T08:42:24Z", "is_private": false, "text": "Steps to reproduce still required."}, {"count": 20, "tags": [], "bug_id": 58646, "attachment_id": null, "id": 188219, "time": "2016-02-04T13:47:57Z", "creator": "mail@majdanski.de", "creation_time": "2016-02-04T13:47:57Z", "is_private": false, "text": "Hi everybody, we have exactly the same problem. We get every few hours \"Error parsing HTTP request header\" and then we must restart tomcat. Does anyone have a solution for this problem? Thanks."}, {"count": 21, "tags": [], "creator": "anmalyshev@mail.ru", "attachment_id": null, "is_private": false, "id": 188442, "time": "2016-02-15T15:16:33Z", "bug_id": 58646, "creation_time": "2016-02-15T15:16:33Z", "text": "Ok, I added some additional logs and here is what we have:\n\nFeb 15, 2016 MSK 05:34:24.157 PM http-nio-135.60.87.68-443-exec-34 org.apache.coyote.http11.Http11NioProtocol\nINFO: Starting to process processor with request 500. Procassor source: connections map.Calling method is:\norg.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:634)\n\nFeb 15, 2016 MSK 05:34:24.157 PM http-nio-135.60.87.68-443-exec-34 org.apache.coyote.http11.Http11NioProcessor\nINFO: Processing GET request to /viewer/wcc/app.swf.\n\nFeb 15, 2016 MSK 05:34:24.157 PM http-nio-135.60.87.68-443-exec-34 com.avaya.acc.base.tomcat.MCPTomcatAccessLogValve\nINFO: 135.60.87.91 - - [15/Feb/2016:17:34:23 +0300] \"GET /viewer/wcc/app.swf HTTP/1.1\" 200 -\n\n...\n\nFeb 15, 2016 MSK 05:34:24.170 PM http-nio-135.60.87.68-443-exec-34 org.apache.coyote.AbstractProtocol$RecycledProcessors\nINFO: Pushing processor with request 500 to RecycledProcessors. Calling method is:\norg.apache.coyote.AbstractProtocol$RecycledProcessors.push(AbstractProtocol.java:906)\n\nFeb 15, 2016 MSK 05:34:24.170 PM http-nio-135.60.87.68-443-exec-34 org.apache.coyote.AbstractProtocol$RecycledProcessors\nINFO: Pushing processor with request 500 to RecycledProcessors. Calling method is:\norg.apache.coyote.AbstractProtocol$RecycledProcessors.push(AbstractProtocol.java:906)\n\nAs you can see, thread just processes standard static content request and after request is processed it pushes processor 2 times to the RecycledProcessors map..\n\nNote that I have incorrect calling method log in last two lines, so it's hard to say which exact places into AbstractProtocol.AbstractConnectionHandler.process() was called (it has 5 occurences of release() method call, which puts processor back to RecycledProcessors map). In case of upgrade request processing it's theoretically possible (first call in \"if (state == SocketState.UPGRADING) {...}\" block and second in the end of process() method if IOExcepton is thrown after release is called first time). But, having in mind I've not used websocket at all during this test, I'm not sure how it could call release twise.\n\nI'll correct last log and try to reproduce again."}, {"count": 22, "tags": [], "bug_id": 58646, "text": "Oops.\n\nSome incorrect info in last comment. It turns out I used both comet(implemented using async servlets) and websocket. So call path with \"if (state == SocketState.UPGRADING) {...}\" and subsequent IOException is possible. Will try to reproduce it.", "id": 188443, "time": "2016-02-15T15:48:45Z", "creator": "anmalyshev@mail.ru", "creation_time": "2016-02-15T15:48:45Z", "is_private": false, "attachment_id": null}, {"count": 23, "tags": [], "bug_id": 58646, "attachment_id": null, "id": 188459, "time": "2016-02-15T21:25:36Z", "creator": "markt@apache.org", "creation_time": "2016-02-15T21:25:36Z", "is_private": false, "text": "Something to try (that would address the Upgrade + IOE issue above and any similar issues) is to add this:\nconnections.remove(socket.getSocket());\n\nto the start of:\nHttp11NioProtocol.Http11ConnectionHandler.release(SocketWrapper,Processor,boolean,boolean)"}, {"text": "I got some new info here.\n\n1) This issue is hardly reproducible with the logs because logging for each request highly slows down my server and request rate drops. I removed logs for all request and now I'm only logging the scenario when processor is recycled and put in RecycledProcessors map twice. And issue is reproducible very good with such logging.\n\n2) Here is what happens. In very rare cases processor is recycled twice. Here is the info on how it gets recycled twice (note that I applied change from Mark's last comment - It does not help at all). I'm remembering stack trace line and thread name from last recycle call and if recycle is called once again, I'm logging it out:\n\nFeb 17, 2016 MSK 03:49:37.417 PM http-nio-135.60.87.68-443-exec-50 org.apache.coyote.http11.AbstractHttp11Processor\nSEVERE: Exception!!! Processor with request 676 to URI /viewer/wcc/app.swf is recycled twice.\n\nFirst Calling method is: org.apache.tomcat.util.net.NioEndpoint$Poller.cancelledKey(NioEndpoint.java:953) by thread http-nio-135.60.87.68-443-exec-50\nCurrent Calling method is: org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:785).\n\nand right after that I'm getting usual NPEs and other exceptions related to this processor:\n\nFeb 17, 2016 MSK 03:49:37.425 PM http-nio-135.60.87.68-443-exec-46 org.apache.coyote.http11.Http11NioProcessor\nINFO: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.. Request id: 676\n\njava.lang.IllegalStateException: Unexpected state: headers already parsed. Buffer not recycled?\n\tat org.apache.coyote.http11.AbstractNioInputBuffer.parseHeaders(AbstractNioInputBuffer.java:375)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1034)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:686)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\nFeb 17, 2016 MSK 03:49:37.427 PM http-nio-135.60.87.68-443-exec-46 org.apache.coyote.http11.Http11NioProcessor\nSEVERE: Error finishing response. Request id: 676\n\njava.lang.NullPointerException\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.flushBuffer(InternalNioOutputBuffer.java:234)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.addToBB(InternalNioOutputBuffer.java:189)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.commit(InternalNioOutputBuffer.java:177)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:748)\n\tat org.apache.coyote.Response.action(Response.java:177)\n\tat org.apache.coyote.http11.AbstractOutputBuffer.endRequest(AbstractOutputBuffer.java:369)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.endRequest(AbstractHttp11Processor.java:1818)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1157)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:686)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\nNote that calling thread in both cases was http-nio-135.60.87.68-443-exec-50 - which most probably means it happened during the same AbstractConnectionHandler.process() call. I also noticed that it always happens in the very begining of my traffic run and first time issue happens with the request to static file (/viewer/wcc/app.swf). With the fact that one of the possible call paths to cancelledKey() method is through NioEndpoint.processSendfile() method - my opinion is that it's 90% related to send file fucntionality and special logic for that.", "tags": [], "creator": "anmalyshev@mail.ru", "is_private": false, "count": 24, "id": 188502, "time": "2016-02-17T13:13:54Z", "bug_id": 58646, "creation_time": "2016-02-17T13:13:54Z", "attachment_id": null}, {"count": 25, "tags": [], "bug_id": 58646, "attachment_id": null, "id": 188503, "time": "2016-02-17T13:17:10Z", "creator": "anmalyshev@mail.ru", "creation_time": "2016-02-17T13:17:10Z", "is_private": false, "text": "I will try to remember the whole stack trace of the first call to find out the actual call path - as there are a lot of possible call paths to cancelledKey() method."}, {"count": 26, "tags": [], "creator": "anmalyshev@mail.ru", "attachment_id": null, "id": 188505, "time": "2016-02-17T13:58:20Z", "bug_id": 58646, "creation_time": "2016-02-17T13:58:20Z", "is_private": false, "text": "Ok, I confirmed it's related to send file functionality:\n\nFeb 17, 2016 MSK 04:31:03.077 PM http-nio-135.60.87.68-443-exec-25 org.apache.coyote.http11.AbstractHttp11Processor\nSEVERE: Exception!!! Processor with request 219 to URI /viewer/wcc/app.swf is recycled twice.\nFirst Calling stack trace is by thread http-nio-135.60.87.68-443-exec-25:\njava.lang.Thread.getStackTrace(Thread.java:1552)\norg.apache.coyote.http11.AbstractHttp11Processor.recycle(AbstractHttp11Processor.java:1869)\norg.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.release(Http11NioProtocol.java:199)\norg.apache.tomcat.util.net.NioEndpoint$Poller.cancelledKey(NioEndpoint.java:953)\norg.apache.tomcat.util.net.NioEndpoint$Poller.processSendfile(NioEndpoint.java:1212)\norg.apache.coyote.http11.Http11NioProcessor.breakKeepAliveLoop(Http11NioProcessor.java:285)\norg.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1186)\norg.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:686)\norg.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\norg.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\njava.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\njava.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\norg.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\njava.lang.Thread.run(Thread.java:745)\nCurrent Calling method is: org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:785)\n\nFeb 17, 2016 MSK 04:31:03.099 PM http-nio-135.60.87.68-443-exec-26 org.apache.coyote.http11.Http11NioProcessor\nINFO: Error parsing HTTP request header\n Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.. Request id: 219\n\njava.lang.IllegalStateException: Unexpected state: headers already parsed. Buffer not recycled?\n\tat org.apache.coyote.http11.AbstractNioInputBuffer.parseHeaders(AbstractNioInputBuffer.java:375)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1034)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:686)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\n\nFeb 17, 2016 MSK 04:31:03.100 PM http-nio-135.60.87.68-443-exec-26 org.apache.coyote.http11.Http11NioProcessor\nSEVERE: Error finishing response. Request id: 219\n\njava.lang.NullPointerException\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.flushBuffer(InternalNioOutputBuffer.java:234)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.addToBB(InternalNioOutputBuffer.java:189)\n\tat org.apache.coyote.http11.InternalNioOutputBuffer.commit(InternalNioOutputBuffer.java:177)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:748)\n\tat org.apache.coyote.Response.action(Response.java:177)\n\tat org.apache.coyote.http11.AbstractOutputBuffer.endRequest(AbstractOutputBuffer.java:369)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.endRequest(AbstractHttp11Processor.java:1818)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1157)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:686)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)\n\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:745)\n\nIf additional info is enabled to find the root cause, please let me know."}, {"count": 27, "tags": [], "creator": "markt@apache.org", "attachment_id": null, "is_private": false, "id": 188578, "time": "2016-02-18T16:47:55Z", "bug_id": 58646, "creation_time": "2016-02-18T16:47:55Z", "text": "Thanks for all your research. Once I knew it was sendfile, tracking it down was fairly simple. I have fixed this in 8.0.x for 8.0.33. Please test and feel free to re-open if it isn't fixed."}, {"text": "Fix back-ported to 7.0.x for 7.0.69 onwards and to 6.0.x for 6.0.46 onwards.", "tags": [], "creator": "markt@apache.org", "attachment_id": null, "count": 28, "id": 188582, "time": "2016-02-18T19:17:57Z", "bug_id": 58646, "creation_time": "2016-02-18T19:17:57Z", "is_private": false}, {"count": 29, "tags": [], "bug_id": 58646, "text": "I have encountered a very similar issue on Tomcat 9.0.0.M1: \n- first \"CoyoteAdapter.checkRecycled Encountered a non-recycled request and recycled it forcedly\",\n- then a lof of 400 errors with the tomcat error \"Http11Processor.endRequest Error finishing response java.lang.NullPointerException\"\n\nI would like to switch back to the stable version of Tomcat 8.5.14. However when I look at the changelog, I cannot see that this issue is corrected in the 8.5.X branch (nor in the 9.0.0.X branch).\n\nIs my bug another issue or is this bug not yet backport to this branches?\nThank you for your precisions.", "id": 198541, "time": "2017-04-26T08:45:11Z", "creator": "amanteaux+bzapache@gmail.com", "creation_time": "2017-04-26T08:45:11Z", "is_private": false, "attachment_id": null}, {"count": 30, "tags": [], "bug_id": 58646, "text": "Yes, the original bug is fixed in all 8.5.x and 9.0.x releases. It doesn't appear in the changelog because it was fixed before development of those branches started.", "id": 198542, "time": "2017-04-26T09:09:42Z", "creator": "markt@apache.org", "creation_time": "2017-04-26T09:09:42Z", "is_private": false, "attachment_id": null}]